### PR TITLE
Get enriched Connection information

### PR DESCRIPTION
**Pull request by jjz921024**
Currently, the information returned by `tostring()` of connection is very poor. we can refer the info format of netty channel to represent a connection 

https://github.com/netty/netty/blob/3c65eae052b7f460e054e6ce765cfba7f433ea32/transport/src/main/java/io/netty/channel/AbstractChannel.java#L382-L419

### CODE REVIEW

**Comment by sazzad16** [comment]
Note: Due to unique id, existing usages like `aConnection.toString(otherConnection.toString())` will fail.

**Comment by jjz921024** [comment]
> Note: Due to unique id, existing usages like `aConnection.toString(otherConnection.toString())` will fail.

@sazzad16 Hi, Thanks for pointing out that this will bring breaking change. For compatibility, can we add two new method that provides `remoteAddr` and `localAddr` for user and keep `toString` method. 

I think the address info of `connection` is very useful for user at debug.

**Comment by sazzad16** [comment]
@jjz921024 

> Hi, Thanks for pointing out that this will bring breaking change. For compatibility, can we add two new method that provides `remoteAddr` and `localAddr` for user and keep `toString` method.

That's possible.

You can also put all the necessary information into a single String, just without overriding toString(). This is something I started in #3137 due to #3058 but didn't move forward because of lack of traction.

**Comment by jjz921024** [comment]
I have updated for this feature. Add two new `getRemoteHostAndPort()` and `getLocalHostAndPort()` method at connection.
When socket is connected, they will return remote and local address info respectively.  @sazzad16 please review, thank.

**Comment by sazzad16** [comment]
@jjz921024 Apologies for the dilemma. I prefer your earlier implementation. WDYT about bringing it back; with renaming `toString()` to something like `toIdentityString()`.

**Comment by jjz921024** [comment]
> @jjz921024 Apologies for the dilemma. I prefer your earlier implementation. WDYT about bringing it back; with renaming `toString()` to something else (like `toIdentityString()`).

That's fine. I had reverted the last commit. Please review.

File: src/main/java/redis/clients/jedis/Connection.java [hunk 1]
```diff
@@ -67,9 +70,43 @@ public Connection(final JedisSocketFactory socketFactory, JedisClientConfig clie
     initializeFromClientConfig(clientConfig);
   }
 
-  @Override
-  public String toString() {
-    return "Connection{" + socketFactory + "}";
+  public String toIdentityString() {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Connection.java:73]: Keep the existing one.
```suggestion
  @Override
  public String toString() {
    return "Connection{" + socketFactory + "}";
  }

  public String toIdentityString() {
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/Connection.java [hunk 2]
```diff
@@ -67,9 +70,43 @@ public Connection(final JedisSocketFactory socketFactory, JedisClientConfig clie
...
+    } else {
+      StringBuilder buf = new StringBuilder(16)
+          .append("[id: 0x")
+          .append(id)
+          .append(']');
+      strVal = buf.toString();
+    }
+
+    strValActive = broken;
+    return strVal;
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Connection.java:109]: Include the class name using `getClass()`.

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/Connection.java (hunk 1 above)
Comment (jjz921024) [reply src/main/java/redis/clients/jedis/Connection.java:73]: Ohhh.... Sorry about it. My mistake.

**Comment by jjz921024** [comment]
@sazzad16 Hi, I found there are a NullPointException when invoke `toIdentityString` before connection is connected and had fixed  it. Please review.

**Review by sazzad16 — CHANGES_REQUESTED** [review]
LGTM!
Please add unit test. At least a simple one.

**Comment by jjz921024** [comment]
> LGTM! Please add unit test. At least a simple one.

@sazzad16 I had added a test case. Please review.  :)

**Review by sazzad16 — APPROVED** [review]

**Pull request merged**